tokio-timer/src/timer.rs

```diff
@@ -0,0 +1,992 @@
+/// range of time equal to the wheel at the lower level. At level zero, each
+/// slot represents one millisecond of time.
+///
+/// The wheels are:
+///
+/// * Level 0: 64 x 1 millisecond slots.
+/// * Level 1: 64 x 64 millisecond slots.
+/// * Level 2: 64 x ~4 second slots.
+/// * Level 3: 64 x ~4 minute slots.
+/// * Level 4: 64 x ~4 hour slots.
+/// * Level 5: 64 x ~12 day slots.
+///
+/// [`Sleep`]: ../struct.Sleep.html
+/// [`Interval`]: ../struct.Interval.html
+/// [`Deadline`]: ../struct.Deadline.html
+/// [paper]: http://www.cs.columbia.edu/~nahum/w6998/papers/ton97-timing-wheels.pdf
+#[derive(Debug)]
+pub struct Timer<T, N = SystemNow> {
+    /// Shared state
+    inner: Arc<Inner>,
+
+    /// The number of milliseconds elapsed since the timer started.
+    elapsed: u64,
+
+    /// Timer wheel.
+    ///
+    /// Levels:
+    ///
+    /// * 1 ms slots / 64 ms range
+    /// * 64 ms slots / ~ 4 sec range
+    /// * ~ 4 sec slots / ~ 4 min range
+    /// * ~ 4 min slots / ~ 4 hr range
+    /// * ~ 4 hr slots / ~ 12 day range
+    /// * ~ 12 day slots / ~ 2 yr range
+    levels: Vec<Level>,
+
+    /// Thread parker. The `Timer` park implementation delegates to this.
+    park: T,
+
+    /// Source of "now" instances
+    now: N,
+}
+
+/// Return value from the `turn` method on `Timer`.
+///
+/// Currently this value doesn't actually provide any functionality, but it may
+/// in the future give insight into what happened during `turn`.
+#[derive(Debug)]
+pub struct Turn(());
+
+/// Registration with a timer.
+///
+/// The association between a `Sleep` instance and a timer is done lazily in
+/// `poll`
+#[derive(Debug)]
+pub(crate) struct Registration {
+    entry: Option<Arc<Entry>>,
+}
+
+pub(crate) struct Inner {
+    /// The instant at which the timer started running.
+    start: Instant,
+
+    /// The last published timer `elapsed` value.
+    elapsed: AtomicU64,
+
+    /// Number of active timeouts
+    num: AtomicUsize,
+
+    /// Head of the "process" linked list.
+    process_head: AtomicPtr<Entry>,
+
+    /// Unparks the timer thread.
+    unpark: Box<Unpark>,
+}
+
+#[derive(Debug)]
+struct Entry {
+    /// Timer internals
+    inner: Weak<Inner>,
+
+    /// Task to notify once the deadline is reached.
+    task: AtomicTask,
+
+    /// Tracks the entry state
+    state: AtomicUsize,
+
+    /// Next entry in the "process" linked list.
+    ///
+    /// Represents a strong Arc ref.
+    next_queue: UnsafeCell<*mut Entry>,
+
+    /// `Sleep` deadline
+    deadline: Instant,
+
+    /// Next entry in the State's linked list.
+    ///
+    /// This is only accessed by the timer
+    next_state: UnsafeCell<Option<Arc<Entry>>>,
+
+    /// Previous entry in the State's linked list.
+    ///
+    /// This is only accessed by the timer and is used to unlink a canceled
+    /// entry.
+    ///
+    /// This is a weak reference.
+    prev_state: UnsafeCell<*const Entry>,
+}
+
+struct Level {
```

Comment:
Would you please split these structs into files? LOC of this file is almost 1k

Reply:
All the code in here is pretty interwoven as it is the core timer implementation. I think keeping it together in the same file is easier to reason about as it is one "unit" of abstraction.

Reply:
I ended up splitting it up.